doc: Updates to Storage module readme (#390)
nrfconnect/Asset-Tracker-Template: docs/modules/storage.md

The storage module forwards or stores data from enabled modules. 
It has following two modes:

- Passthrough (default): forward data immediately.
- Buffer: stores data, retrieves through flush or batch.

It is implemented as a small SMF state machine with a parent `RUNNING` state and `PASSTHROUGH/BUFFER` children state. 
Data types are discovered automatically through iterable sections. 
See `storage.c`, `storage.h`, and `Kconfig.storage` for details.

## Architecture

### States

The storage module implements a state machine with the following states and transitions:

![Storage module state diagram](../images/storage_module_state_diagram.svg "Storage module state diagram")

- **RUNNING** (parent): Initializes backend, handles admin commands (`STORAGE_CLEAR`, `STORAGE_FLUSH`, `STORAGE_STATS`).
- **PASSTHROUGH**: Forwards data immediately as `STORAGE_DATA` on `STORAGE_DATA_CHAN`.
- **BUFFER**: Stores data in backend, serves flush and batch requests. Has sub-states `IDLE` and `PIPE_ACTIVE`.

Initial state: `PASSTHROUGH` (default) or `BUFFER` (if `CONFIG_APP_STORAGE_INITIAL_MODE_BUFFER=y`).
Runtime switching: Send `STORAGE_MODE_*_REQUEST` on `STORAGE_CHAN`.

### Backend

> Backends are implemented in the `app/src/modules/storagestorage_backend.h` file and provide `init`, `store`, `peek`, `retrieve`, `count`, and `clear` functionalities.

Reviewer: ```suggestion
Backends implements the API defined in  `app/src/modules/storagestorage_backend.h` file and provide `init`, `store`, `peek`, `retrieve`, `count`, and `clear` functionalities.
```

The default backend is RAM.

### Data flow

> Modules are published to zbus. 

Reviewer: ```suggestion
Data producing modules publish sampled data to their respective zbus channel. 
```

- Per-type slabs: Each enabled data type declares a `k_mem_slab` with `CONFIG_APP_STORAGE_MAX_RECORDS_PER_TYPE` blocks of that type's size.
- RAM backend ring buffers: For each enabled data type, a ring buffer is declared with capacity `sizeof(type) * CONFIG_APP_STORAGE_MAX_RECORDS_PER_TYPE`.
- Message buffers: `struct storage_msg` carries a `buffer[STORAGE_MAX_DATA_SIZE]`, where `STORAGE_MAX_DATA_SIZE` is the max size of any enabled data type. 
  Enabling large types increases this buffer and several temporary buffers.
- Subscriber queue: Size is controlled by system zbus configuration.
- Thread stack: `CONFIG_APP_STORAGE_THREAD_STACK_SIZE`.

#### How to reduce RAM

- Minimize enabled data types

    - Disable modules that you do not forward or store (for example, `CONFIG_APP_LOCATION=n`), which reduces both slabs and RAM backend ring buffers, and shrinks `STORAGE_MAX_DATA_SIZE`.

- Reduce records per type

    - Set `CONFIG_APP_STORAGE_MAX_RECORDS_PER_TYPE=1` when buffering is not needed. 
      This shrinks both the per-type slabs and RAM ring buffers to a single record each.

- Shrink batch pipe buffer

>     - Set the `CONFIG_APP_STORAGE_BATCH_BUFFER_SIZE` Kconfig option to a lowest value. 

Reviewer: ```suggestion
    - Set the `CONFIG_APP_STORAGE_BATCH_BUFFER_SIZE` Kconfig option to a low value. 
```

      The batch pipe is allocated unconditionally, but it is only used in the buffer mode. 
      If you never use the batch or buffer mode, pick a very small value (for example, from 64 to 256 bytes). 

>       If you do use batch, ensure it can hold `sizeof(header) + max_item_size`.

Reviewer: ```suggestion
      If you do use batch, ensure it can hold at least one item of `sizeof(header) + max_item_size`.
```

- Reduce thread and queues

    - Set the `CONFIG_APP_STORAGE_THREAD_STACK_SIZE` Kconfig option to a lower value (for example, from 2048 to 1024), if your application leaves headroom.
    - Reduce relevant zbus queue sizes in system config if traffic allows.

- Remove development features

>     - Disable the `CONFIG_APP_STORAGE_SHELL` and `CONFIG_APP_STORAGE_SHELL_STATS` Kconfig option to trim RAM or code.

Reviewer: ```suggestion
    - Disable the `CONFIG_APP_STORAGE_SHELL` and `CONFIG_APP_STORAGE_SHELL_STATS` Kconfig option to trim RAM and code footprint.
```

# CONFIG_APP_NETWORK=n
```

!!! note "Note"

    - If you later enable buffer or batch, increase the value of the `CONFIG_APP_STORAGE_BATCH_BUFFER_SIZE` Kconfig option, so that at least one header plus the largest item fits.
    - The actual RAM consumed by ring buffers and slabs scales with which data types are enabled and the value of the `CONFIG_APP_STORAGE_MAX_RECORDS_PER_TYPE` Kconfig option.

## Messages

The storage module communicates through two zbus channels: `STORAGE_CHAN` and `STORAGE_DATA_CHAN`. 
All message types are defined in the `storage.h` file.

### Input Messages (Commands)

**Mode Control:**

- **`STORAGE_MODE_PASSTHROUGH_REQUEST`**: Request to switch from `buffer` mode to `passthrough` mode. 
  On success, the module replies with `STORAGE_MODE_PASSTHROUGH`. 

>   If the change is unsafe (for example, batch session active), the module replies with `STORAGE_MODE_CHANGE_REJECTED`.

Reviewer: ```suggestion
  If the change is declined (for example, batch session active), the module replies with `STORAGE_MODE_CHANGE_REJECTED`.
```

  On success, the module replies with `STORAGE_MODE_BUFFER`. 
  If the change is unsafe, the module replies with `STORAGE_MODE_CHANGE_REJECTED`.

**Data Operations (handled by parent RUNNING state):**

- **`STORAGE_FLUSH`**: Flushes stored data one item at a time as individual `STORAGE_DATA` messages. 
  Data is sent in FIFO order per type. Available in both operational modes.

- **`STORAGE_BATCH_REQUEST`**: Requests access to stored data through batch interface. 
  Responds with `STORAGE_BATCH_AVAILABLE`, `STORAGE_BATCH_EMPTY`, `STORAGE_BATCH_BUSY`, or `STORAGE_BATCH_ERROR`. 
  Available in both operational modes.

- **`STORAGE_CLEAR`**: Clears all stored data from the backend. 
  Available in both operational modes.

**Diagnostics (handled by parent RUNNING state):**

- **`STORAGE_STATS`** : Requests storage statistics (requires `CONFIG_APP_STORAGE_SHELL_STATS`). 
  Statistics are logged to the console. 
  Available in both operational modes.

### Output Messages (Responses)

**Mode confirmation:**

- **`STORAGE_MODE_PASSTHROUGH`**: Mode change to passthrough confirmed.

- **`STORAGE_MODE_BUFFER`**: Mode change to buffer confirmed.

> - **`STORAGE_MODE_CHANGE_REJECTED`**: Mode change rejected due to safety constraints. 

Reviewer: ```suggestion
- **`STORAGE_MODE_CHANGE_REJECTED`**: Mode change rejected. 
```

## Adding Backends

> 1. Implement `struct storage_backend` (see `storage_backend.h`).

Reviewer 2: Should this be 1. -  3. ?
